source/MaterialXCore/Node.cpp

```diff
@@ -675,7 +681,19 @@ void NodeGraph::modifyInterfaceName(const string& inputPath, const string& inter
 
 NodeDefPtr NodeGraph::getNodeDef() const
 {
-    return resolveRootNameReference<NodeDef>(getNodeDefString());
+    NodeDefPtr nodedef = resolveRootNameReference<NodeDef>(getNodeDefString());
+    // If not directly defined look for an implementation which has a nodedef association
+    if (!nodedef)
+    {
+        for (auto impl : getDocument()->getImplementations())
```

Comment:
This requires a traversal now. To see if this causes performance issue. If so perhaps consider caching the nodegraph->nodedef associations.

Reply:
As an initial commit I suppose this is ok. But I wonder if we should remove this method `NodeGraph::getNodeDef()` altogether, now that this association has been removed. And instead update any part of codegen that uses this method.

Reply:
This can be called outside of codegen to allow bidirectional lookups: (nodegraph<->nodedef, equivalent to node <-> nodedef ) so don't think we should break this this API.  

Guess this can be visited once what to do about upgrading is determined, as I believe no deprecation logic has been agreed upon yet.

Reply:
Yes, as agreed on offline we should consider this an additive change and keep the old syntax and API's for now.